lib/ddtrace/contrib/active_record/patcher.rb

```diff
@@ -0,0 +1,88 @@
+module Datadog
+  module Contrib
+    module ActiveRecord
+      # Patcher enables patching of 'active_record' module.
+      # This is used in monkey.rb to manually apply patches
+      module Patcher
+        @patched = false
+
+        module_function
+
+        # patched? tells whether patch has been successfully applied
+        def patched?
+          @patched
+        end
+
+        def patch
+          if !@patched && defined?(::ActiveRecord)
+            begin
+              require 'ddtrace/contrib/rails/utils'
+              require 'ddtrace/ext/sql'
+
+              patch_active_record()
+
+              # XXX This module currently only supports sinatra (see
+              # self.datadog_trace).
+              tracer().set_service_info(database_service(), 'sinatra',
+                                        Datadog::Ext::AppTypes::DB)
+
+              @patched = true
+            rescue => e
+              Datadog::Tracer.log.error("Unable to apply Active Record integration: #{e}")
+            end
+          end
+
+          @patched
+        end
+
+        def patch_active_record
+          # subscribe when the active record query has been processed
+          ::ActiveSupport::Notifications.subscribe('sql.active_record') do |*args|
+            sql(*args)
+          end
+        end
+
+        def self.datadog_trace
+          # TODO: Consider using patcher for Rails as well.
+          # @tracer ||= defined?(::Rails) && ::Rails.configuration.datadog_trace
+          @datadog_trace ||= defined?(::Sinatra) && ::Sinatra::Application.settings.datadog_tracer.cfg
+        end
+
+        def self.adapter_name
+          @adapter_name ||= Datadog::Contrib::Rails::Utils.normalize_vendor(
+            ::ActiveRecord::Base.connection_config[:adapter]
+          )
+        end
+
+        def self.tracer
+          @tracer ||= datadog_trace.fetch(:tracer)
+        end
+
+        def self.database_service
+          @database_service ||= datadog_trace.fetch(:default_database_service, adapter_name)
```

Comment:
OK I obviously was wrong in hinting we could call set_service_info in patcher, as the database_service is not known so early in the exec flow. Could we technically do it here? It seems that the case `@database_service` is nil is quite rare (only the first time) and at least here, we now datadog_trace is defined and won't bail out with an error when calling fetch?